qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+        Because the order of elements in the block is instrumental to scheduling, instructions can't
+        be removed if the block is not scheduled.
+
+        Args:
+            instruction: The instruction to be removed.
+
+        Raises:
+            PulseError: If the ``IrBlock`` is not scheduled.
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not remove instruction from unscheduled IrBlock")
+        if instruction in self._elements:
+            self._elements.remove(instruction)
+
+    def filter_and_remove_instructions(
+        self, filter_function: Callable, recursive: bool = False
+    ) -> None:
+        """Filter and remove instructions from ``IrBlock`` (in place)
+
+        Instructions for which ``filter_function`` return ``False`` are removed from the block.
+        After the application, the order of elements will be changed, such that nested blocks
+        will be first, and instructions last.
+
+        Because the order of elements in the block is instrumental to scheduling, instructions can't
+        be removed if the block is not scheduled.
+
+        Args:
+            filter_function: A Callable which takes as argument an ``IrInstruction`` and returns
+                ``False`` to remove the instruction, and ``True`` to keep it.
+            recursive: If ``True``, applies the filter to nested ``IrBlock``s. Default value - ``False``.
+
+        Raises:
+            PulseError: If the ``IrBlock`` is not scheduled.
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not remove instructions from unscheduled IrBlock")
+
+        blocks = [element for element in self.elements if isinstance(element, IrBlock)]
+        if recursive:
+            for block in blocks:
+                block.filter_and_remove_instructions(filter_function, True)
+        instructions = [element for element in self.elements if isinstance(element, IrInstruction)]
+        instructions = list(filter(filter_function, instructions))
+        self._elements = blocks + instructions
+
+    def __eq__(self, other: "IrBlock") -> bool:
+        """Return True iff self and other are equal
+        Specifically, iff all of their properties are identical.
+
+        Args:
+            other: The IrBlock to compare to this one.
+
+        Returns:
+            True iff equal.
+        """
+        if (
+            type(self) is not type(self)
+            or self._alignment != other._alignment
+            or len(self._elements) != len(other._elements)
+        ):
+            return False
+        for element_self, element_other in zip(self._elements, other._elements):
+            if element_other != element_self:
+                return False
+
+        return True
+
+    def __len__(self) -> int:
+        """Return the length of the IR, defined as the number of elements in it"""
+        return len(self._elements)
+
+    def __repr__(self) -> str:
```

Comment:
In my experience repr for sequence (e.g. ScheduleBlock) is not really useful, because it's super lengthy. If we prefer LLVM IR like, probably we can support `.dump` method that generates parsable code. This also helps debugging (unittest).

Reply:
(note that I don't suggest implementing this in this PR, since you need to start from defining AST...)